client/packages/programs/src/JsonForms/useJsonForms.tsx

```diff
@@ -121,6 +123,7 @@ export const useJsonForms = <R,>(
   }, [loadedData]);
   // current modified data
   const [data, setData] = useState<JsonData | undefined>();
+  const submitActions = useRef<Record<string, () => void>>({});
```

Comment:
Using `useRef` rather than `useState` to reduce re-renders (the UI doesn't need updating in response to changes of this object)

Reply:
![image](https://github.com/user-attachments/assets/a3aa7943-598f-4e86-b695-514f8f990308)

That's one of the intended usages for useRef (along with direct DOM manipulation):

https://react.dev/reference/react/useRef